protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+
+Outside of interop, deployment also seems increasingly complex and opaque, while it does not have to be,
+due to the same configuration and composability troubles.
+
+## DeployConfig complexity
+
+The `DeployConfig` definition, as per the Go code, consists of 102 configurable variables.
+Each one of these may affect the chain deployment at any point during deployment
+due to lack of grouping and encapsulation.
+
+Within `op-chain-ops`, we have started introducing some grouping of features:
+[PR 11189](https://github.com/ethereum-optimism/optimism/pull/11189).
+
+But this is still limited to the Go side, and the complexity still remains on the solidity deploy scripts.
+
+## Stages of deployment
+
+A L2 chain deployment runs through the following steps:
+1. Configure `DeployConfig` (many choices).
+2. Create L1 chain, if developing.
+3. Deploy superchain-wide contracts to L1 (including a `ProxyAdmin` that should not be shared with step 4)
+4. Deploy L2-instance-specific contracts to L1.
+5. Collect deployments in one file, where there is 1 of everything (breaks interop).
+6. Run the L2-genesis script, depending on L1 deployments and the deploy config.
+7. Run tools to turn the L2 genesis state into a chain genesis and rollup configuration.
+
+The configuration of step (1) forces how all the following steps go,
+making it impractical to repeat step (2) per L2 chain.
+
+The L1 deployments artifact that bundles superchain and L2-instance-specific
+deployments also prevents multiple L2s from running.
+
+The main goal is to unbundle these steps, and make them customizable and composable for better test-coverage,
+including Interop parameters specifically.
+
+## OPStackManager
+
+Specs: https://specs.optimism.io/experimental/op-stack-manager.html
+
+Older implementatipn draft: https://github.com/ethereum-optimism/optimism/pull/9985
+
+The OP-Stack manager focuses on making the deployment of the L1 contracts of an L2 chain deterministic,
+by deploying proxies to deterministic addresses (salt = chainID),
+and coupling to superchain-provided implementation contracts.
+
+This implies a different L1-deployment flow than we have today,
+and limits it to just the L2-instance specific contracts.
+A deployment phase we also want to isolate the configuration and execution of.
+
+Note that the draft sources configuration inputs explicitly, rather than coupling to the (arguably bloated) 
+deploy config.
+
+## Deployer requirements
+
+### Op-e2e needs
+
+As an op-e2e test suite, I need to:
+- Customize the deploy-config deeply, including L1 parameters and upgrade parameters, to cover all functionality.
+- Create multiple L2s with a shared L1 (to go from isolated L2s to interoperating L2s through interop hardfork).
+- Create multiple L2s with a shared L1 and shared superchain (to test interop without emulating hardfork first).
+
+### Devnet needs
+
+As a devnet user, I need to:
+- Modify the chain template, for experimentation when building new features.
```

Comment:
What is the chain template, is that what is currently the deploy config?

Reply:
There is a `devnetL1_template.json` file, that turns into a deploy-config, as part of the devnet. But it's a singular template, not possible to change per test.

Reply:
I see, thanls